Add functional testing

betaredex opened the pull request:
I have a basic test script which tests the following actions: offer creation, offer listing, offer claiming, lease listing, lease deletion, and offer deletion. I'm not sure how I should set up error handling, or how I should organize the tests into different scripts. Most tests will rely on offer creation and deletion working at least, so if I make multiple scripts, those would probably need to be in all of them.

DanNiESh commented:
I think you could create a setup.sh that does offer_creation and a cleanup.sh that deletes the offer and related leases. Then include them in your test scripts.

DanNiESh commented:
We usually squash commits into one each PR. Here is the example: https://www.internalpointers.com/post/squash-commits-into-one-git

betaredex commented:
Additions:
Added setup and cleanup scripts
Made error handling more compact in the offer test script
Made a failure test script for time conflicts with offer creation

The failure test script is a bit messy at the moment, and I want to think about how to name things better

larsks commented on tests/functional/test-offer:
```diff
@@ -0,0 +1,58 @@
+. ~/test-projects/test1-subproject-openrc
+openstack esi offer list -f json > $tmpfile \
+  || { ec=$?; echo "ERROR: failed to list offers" >&2; exit $ec; }
+cat $tmpfile | jq -r ".[].UUID" | grep $uuid \
```
You don't need to use `cat` here. You can write:

```
jq -r ".[].UUID" $tmpfile | grep $uuid```

larsks commented on tests/functional/test-offer:
```diff
@@ -0,0 +1,58 @@
+openstack esi lease list -f json > $tmpfile \
+  || { ec=$?; echo "ERROR: failed to list leases" >&2; exit $ec; }
+
+cat $tmpfile | jq -r ".[].UUID" | grep $uuid \
```
Don't need `cat` here.

larsks commented on tests/functional/test-offer-fail:
```diff
@@ -0,0 +1,41 @@
+ec=$?
+expected_error='Time conflict for dummy_node 1001.'
+
+if ! cat $errfile | grep -q "$expected_error"; then
```
You don't need `cat` here; you can write:

```
if ! grep -q "$expected_error" $errfile; then
```

larsks commented on tests/functional/cleanup:
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+. ~/test-projects/test1-openrc
```
Consider making the path to the rc file a variable (maybe with a default value), rather than requiring a specific file structure in the home directory. Something like:

```
. ${ESI_TEST_CREDENTIALS:-./test-openrc}
```

This would default to a file named `test-openrc` in the current directory if `$ESI_TEST_CREDENTIALS` is unset.

You may also want to have a sane error message if the credentials file is missing.

DanNiESh commented on tests/functional/cleanup:
```diff
@@ -0,0 +1,10 @@
+test_offer_uuid=$1
+
+echo "OFFER DELETE TEST"
+. "$ESI_TEST_CREDENTIALS/test1-openrc"
```
I think @larsks means $ESI_TEST_CREDENTIALS would be the path to rc file.  For example,  set ```ESI_TEST_CREDENTIALS = ~/test-projects/test2-openrc``` at the beginning, then ```. ${ESI_TEST_CREDENTIALS:-./test1-openrc}```, it will source test2 project. But if you don't set ESI_TEST_CREDENTIALS at first, it will default to test1-openrc.

betaredex replied:
I understood that, but I need to use different test projects in the same script, so I can't use one variable for all of them.

DanNiESh commented on tests/functional/setup:
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+. "$ESI_TEST_CREDENTIALS/test1-openrc"
```
same

DanNiESh commented on tests/functional/test-offer:
```diff
@@ -0,0 +1,58 @@
+
+echo "OFFER LIST TEST"
+
+. "$ESI_TEST_CREDENTIALS/test1-subproject-openrc"
```
same

DanNiESh commented on tests/functional/test-offer:
```diff
@@ -0,0 +1,58 @@
+
+echo "LEASE LIST TEST"
+
+. "$ESI_TEST_CREDENTIALS/test1-openrc"
```
same

DanNiESh commented on tests/functional/test-offer-fail:
```diff
@@ -0,0 +1,70 @@
+
+echo "OFFER CREATE TIME CONFLICT TEST"
+
+. "$ESI_TEST_CREDENTIALS/test1-openrc"
```
same

DanNiESh commented on tests/functional/test-offer-fail:
```diff
@@ -0,0 +1,70 @@
+
+echo "OFFER CLAIM TEST"
+
+. "$ESI_TEST_CREDENTIALS/test1-subproject-openrc"
```
^^

DanNiESh commented on tests/functional/test-offer-fail:
```diff
@@ -0,0 +1,70 @@
+
+echo "OFFER CLAIM TIME CONFLICT TEST"
+
+. "$ESI_TEST_CREDENTIALS/test1-subproject-1-openrc"
```
^^ same here

DanNiESh commented on tests/functional/test-offer-fail:
```diff
@@ -0,0 +1,70 @@
+
+# Setting up a claim is needed for the claim time conflict test
+
+echo "OFFER CLAIM TEST"
```
You can other claim_offer_fail cases in this script such as lessee outside test1 can't view or claim the offer

tzumainn commented:
Hi! I think this is a good start. Some quick suggestions:

* I think it's typical for bash scripts to be suffixed with ".sh"
* For clarity, I'd call ESI_TEST_CREDENTIALS ESI_TEST_CREDENTIALS_DIR
* I'd suggest putting all hardcoded values as variables at the top of the script. You already do this with lessee_uuid, but I'd add in ESI_TEST_CREDENTIALS as well as specific rc files such as 'test1-openrc'. If someone else wants to run these scripts in a different environment, we want to make it easy for them to know what environment variables they might want to change. In the longer run, I can imagine moving those hardcoded values into some sort of configuration file that all the scripts read - but that can be done later.

larsks commented:
Actually, instead of requiring a bunch of different rc-style credentials files, how about using a *single* [clouds.yaml](https://docs.openstack.org/python-openstackclient/pike/configuration/index.html#clouds-yaml) format file?  Then you can just set `--os-cloud <cloudname>` to pick which credentials you want.  E.g., given something like:

```
clouds:
  test1:
    auth:
      username: test1
      project_name: test1
      password: "secret"
      auth_url: "http://192.168.1.2:5000"
      user_domain_name: Default
      project_domain_name: Default
    identity_api_version: 3

  test1-subproject:
    auth:
      username: test1-subproject
      project_name: test1-subproject
      password: "another secret"
      auth_url: "http://192.168.1.2:5000"
      user_domain_name: Default
      project_domain_name: Default
    identity_api_version: 3
```

You could then run:

```
openstack --os-cloud test1 offer create ...
```

Or :

```
openstack --os-cloud test1-subproject offer create...
```


Etc. This lets you keep the entire credential setup in a single file.
OpenStack will look for `clouds.yaml` in:

- The current directory
- `$HOME/.config/openstack/`
- `/etc/openstack`

tzumainn commented on tests/functional/setup.sh:
```diff
@@ -0,0 +1,31 @@
+lessee_uuid_option=''
+# Check if the lessee uuid was passed
+if ! [[ -z ${4+x} ]]; then
+  lessee_uuid_option=" --lessee-id $4"
```
Note that this is now just '--lessee' instead of '--lessee-id'

tzumainn commented on tests/functional/test-offer-fail.sh:
```diff
@@ -0,0 +1,93 @@
+openstack --os-cloud test1 esi offer create \
+  --start-time $start \
+  --end-time $end \
+  --lessee-id $lessee_uuid \
```
Ditto

tzumainn commented on tests/functional/test-offer-fail.sh:
```diff
@@ -0,0 +1,93 @@
+  --start-time $start \
+  --end-time $end \
+  --resource-type dummy_node 1002 \
+  --lessee-id $lessee_uuid \
```
Ditto

tzumainn reviewed (commented):
Just a note about some CLI changes

betaredex commented:
Fixed the lessee argument thing

tzumainn commented on tests/functional/setup.sh:
```diff
@@ -0,0 +1,31 @@
+openstack --os-cloud test1 esi offer create \
+  --start-time $start \
+  --end-time $end \
+  --resource-type dummy_node 1001\
```
'1001' should be a variable that you set above

tzumainn commented on tests/functional/test-offer-fail.sh:
```diff
@@ -0,0 +1,93 @@
+. $tmpfile
+
+test_offer_uuid=$uuid
+trap "rm -f $tmpfile $errfile; ./cleanup $test_offer_uuid" EXIT  # Run cleanup if the script exits
```
cleanup.sh

tzumainn commented on tests/functional/test-offer-fail.sh:
```diff
@@ -0,0 +1,93 @@
+end=$(date -d "+5 days" +%Y-%m-%d)
+lessee_uuid=39b16f5117c24072844d55f8d20cea5d
+
+./setup $tmpfile $start $end
```
setup.sh

tzumainn commented on tests/functional/test-offer-fail.sh:
```diff
@@ -0,0 +1,93 @@
+    echo "ERROR: succeeded creating offer despite time conflict" >&2
+    . $tmpfile
+    echo "Deleting time conflict test offer..."
+    ./cleanup $uuid
```
cleanup.sh

tzumainn commented on tests/functional/test-offer.sh:
```diff
@@ -0,0 +1,56 @@
+end=$(date -d "+5 days" +%Y-%m-%d)
+lessee_uuid=39b16f5117c24072844d55f8d20cea5d
+
+./setup $tmpfile $start $end $lessee_uuid
```
setup.sh

tzumainn commented on tests/functional/test-offer.sh:
```diff
@@ -0,0 +1,56 @@
+. $tmpfile
+
+test_offer_uuid=$uuid                               # The uuid variable will be overwritten
+trap "rm -f $tmpfile; ./cleanup $test_offer_uuid" EXIT   # Run cleanup if the script exits
```
cleanup.sh

tzumainn reviewed (changes requested):
Hi! This is pretty good; I ran the tests and they worked great. I think the code could probably be genericized slightly - but bash is weird and complex and I'm fine leaving that update for a future patch.

I've left some inline comments for minor issues; a few more suggestions:

* I think setup.sh and cleanup.sh should create/delete the dummy node with a value based on something like the timestamp; that would prevent conflicts when two people are running the scripts
* I'd also add a README to the directory explaining the expectations in regards to clouds.yaml, especially what clouds are expected in that file. Basically imagine someone new installs ESI and wants to run these tests; what do they need to know in order to do so?
* I'd also suggest adding big comments before each test, just to make it easy for someone looking at the script to break apart each section of the script. Something like:

```
##########################
## LEASE CREATION  TEST ##
##########################
```

DanNiESh commented on tests/functional/test-offer-fail.sh:
```diff
@@ -0,0 +1,93 @@
+errfile=$(mktemp ./errXXXXXX)
+start=$(date +%Y-%m-%d)
+end=$(date -d "+5 days" +%Y-%m-%d)
+lessee_uuid=39b16f5117c24072844d55f8d20cea5d
```
You may want to add a comment to indicate this is test1-subproject

tzumainn replied:
Ah, good point! And now that I think about it, you should now be able to specify this by name, and just do something like
lessee=test1-subproject

betaredex replied:
Didn't realize, I'll change it to that then

DanNiESh commented on tests/functional/test-offer.sh:
```diff
@@ -0,0 +1,56 @@
+tmpfile=$(mktemp ./leaseXXXXXX)
+start=$(date +%Y-%m-%d)
+end=$(date -d "+5 days" +%Y-%m-%d)
+lessee_uuid=39b16f5117c24072844d55f8d20cea5d
```
same here: #test1-subproject

DanNiESh reviewed (commented):
Looks pretty good! And Mainn has already summarized the issues very well. I just left some comments on a minor issue.

betaredex commented:
Added two new tests, for an invalid time frame error and for an invalid offer uuid error
Commented out the invalid user test, going to move it to another script because it takes so much extra setup
Renamed all the script calls in the code to use the new filenames
Set up automatic node creation with mktemp

betaredex commented:
Used the lessee name instead of the uuid

betaredex commented:
Recategorized tests into whether they need a lessee id to be specified when creating an offer. The tests where it doesn't matter are placed somewhat arbitrarily, so it's not the best solution, but it's the most efficient one for my purposes. Specifically, almost all of the negative tests are in the script where a lessee id is not specified, but many could be moved. Added comments to clearly show where each test starts and ends, and on what each of them does, as well as a few additional extraneous comments. Moved the node creation to the setup script.

tzumainn commented on tests/functional/test-lessee-id.sh:
```diff
@@ -0,0 +1,114 @@
+# The uuid variable will be overwritten, so we save it here
+test_offer_uuid=$uuid
+# Run cleanup if the script exits, contains offer delete test
+trap "./cleanup.sh $test_offer_uuid; rm -f $tmpfile $errfile $nodefile" EXIT 
```
Is it possible to delete these files as part of the cleanup script?

betaredex replied:
Yeah, I could do that. I didn't want to bother writing it because it's actually more code than this to pass the filename variables in there, but if you think it'd be cleaner that way I'm happy to.

tzumainn replied:
Okay, looking at the code more closely - I think tmpfile and errfile could stay within the test script, since the scripts actively use the file. I'm wondering if you could move the temp node creation into the setup script, set the node_uuid there, and then delete the node file in the cleanup script without ever passing the nodefile out.

betaredex replied:
Maybe, but with the way mktemp works, the node id is defined as part of the node creation. If I had it in setup.sh, I would have to pass it out somehow.

tzumainn replied:
But aren't you passing out a bunch of information from setup.sh anyway? Is it possible to add the node id there?

betaredex replied:
No, setup.sh just edits the tmpfile with the output from the create offer command. I could append the node_id there though I guess.

tzumainn replied:
Yep, let's do that then! And if you do that - maybe add the tmpfile and errfile as well.

betaredex replied:
I can't add tmpfile, since that's how I'm passing the variables, so the test script needs to know where it is. I can add errfile and nodefile though.

tzumainn reviewed (commented):
Looks pretty good overall! One minor question inline.

betaredex commented:
Added a simple readme (which will probably change if I ever try to make the setup easier)

DanNiESh commented on tests/functional/test-no-lessee-id.sh:
```diff
@@ -0,0 +1,141 @@
+. $tmpfile
+
+echo "OFFER CLAIM TEST SUCCEEDED"
+
```
I think you could also create another offer_claim_test that another lessee can claim this offer for a different time range.

betaredex replied:
You mean only claiming it for part of the time range instead of the whole offer range?

larsks replied:
I suggest focusing on fixing the tests already in this PR, and moving any requests for additional tests to a new issue (and a future PR).

DanNiESh replied:
> You mean only claiming it for part of the time range instead of the whole offer range?

yeah, that's what I mean. But you could ignore this new test for now as lars suggested.

tzumainn replied:
@DanNiESh could you check the test spreadsheet to see if the test you suggested is in there, and add it in if it's not? That's probably the best way to ensure a followup.

DanNiESh replied:
Okay, @betaredex, I added a comment in the spreadsheet about it.

betaredex commented:
Moved the errfile and nodefile creation to the setup script and used the tmpfile to pass the filenames back to the test script

tzumainn commented:
This looks good to me! @DanNiESh any other comments?

DanNiESh commented:
> This looks good to me! @DanNiESh any other comments?

Looks good to me!
